### PR TITLE
Optimizer Unit Testing

### PR DESCRIPTION
**Pull request by HangenYuu**
## Description

Added unit testing for subsequent optimizers.

## Related Issue

- [ ] Closes
- [x] Related to #7 #8 #9

### CODE REVIEW

**Comment by HangenYuu** [comment]
Hi @jessegrabowski, I am drafting a PR for unit testing of optimizer with optax. For PyTorch's [practice](https://github.com/pytorch/pytorch/blob/main/test/test_optim.py), they test for 2 things:
1. Functionality: The optimizer indeed makes updates to parameters that make the loss function go down. This is already confirmed for your implementations and my new ones of AdamW and Adadelta.
2. Implementation: For this one, the idea is having a case where the output is deterministic i.e, for the nth optimizing step, for input of this particular random seed, this optimizer should give this output. This one is harder, I am still experimenting with it. One thing so far is using the same input, same model with parameters all initialized to 0, PyTorch and Optax give different loss but roughly same parameters. Maybe I will use **parameters of the model at an optimizing step** for comparison.

A minor thing: Can you check and close PRs #8 and #9? I want AdamW and/or Adadelta to get merged to do testing on them too.

**Comment by jessegrabowski** [comment]
We need actual unittests on #8 and #9 before we can close them. This notebook is a great starting point, we just need to figure out how to write it as a pytest test file. I can follow your start and try to give some help in the next couple days, im just super busy this week

**Comment by jessegrabowski** [comment]
Sorry for letting this languish. I am getting back around to this project, will look at this ASAP.

File: tests/test_optimizers.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+model.initalize_weights()
+
+
+def test_adam():
+    optim = Adam(model, loss_fn, ndim_out=2, learning_rate=1e-3)
+
+    n_obs = X_normed.shape[0]
+    cutpoints = np.arange(0, n_obs, 1000).tolist()
+    cutpoints += [n_obs]
+    batch_slices = list(zip(cutpoints[:-1], cutpoints[1:]))
```
Comment (jessegrabowski) [inline tests/test_optimizers.py:64]: We can write a helper to generate `batch_slices`, because we're going to need it in every test

File: tests/test_optimizers.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    y_hat = prediction_network(X_in)
+    model = Model(X_in, y_hat)
+    return model
+
+
+X_normed, y_onehot = generate_example_data()
+model = create_simple_model()
+loss_fn = CrossEntropy(expect_onehot_labels=True, expect_logits=True, reduction="mean")
+model.initalize_weights()
```
Comment (jessegrabowski) [inline tests/test_optimizers.py:55]: Put this in a fixture with `scope="session"`

File: tests/test_optimizers.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def generate_example_data():
+    X, y = load_digits(return_X_y=True)
+    y_onehot = OneHotEncoder().fit_transform(y[:, None]).toarray()
+    X_normed = MinMaxScaler().fit_transform(X)
+    return X_normed, y_onehot
+
+
+def create_simple_model():
```
Comment (jessegrabowski) [inline tests/test_optimizers.py:37]:
```suggestion
def create_mlp_classifier():
```

File: tests/test_optimizers.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    for _ in range(n_epochs):
+        all_idx = np.arange(n_obs)
+        np.random.shuffle(all_idx)
+        y_epoch = y_onehot[all_idx, :]
+        X_epoch = X_normed[all_idx, :]
+        for start, stop in batch_slices:
+            idx = slice(start, stop)
+            loss = optim.step(X_epoch[idx], y_epoch[idx])
+            loss_history.append(loss)
```
Comment (jessegrabowski) [inline tests/test_optimizers.py:76]: small optimization: we only need the first and last loss, so no need to store all the intermediate values. Just compute the loss once before starting the training loop, then compare it to whatever is in the `loss` variable after the loop.

File: tests/test_optimizers.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+def test_adam():
+    optim = Adam(model, loss_fn, ndim_out=2, learning_rate=1e-3)
+
+    n_obs = X_normed.shape[0]
+    cutpoints = np.arange(0, n_obs, 1000).tolist()
+    cutpoints += [n_obs]
+    batch_slices = list(zip(cutpoints[:-1], cutpoints[1:]))
+    loss_history = []
+    n_epochs = 10
```
Comment (jessegrabowski) [inline tests/test_optimizers.py:66]:
```suggestion
    n_epochs = 1
```

This is fine if all we're doing is making sure the loss is going down

**Review by jessegrabowski — CHANGES_REQUESTED** [review]
This is really great, again I'm very sorry I let this lie idle. Hope you're still keen to push it forward!

**Comment by jessegrabowski** [comment]
Thanks for this Hang! Only 6 months from opening to merging, a new world record
